internal/handlers/common/projection.go

```diff
@@ -85,34 +108,114 @@ func ProjectDocuments(docs []*types.Document, projection *types.Document) error
-			v, ok := projectionMap[k]
-			if !ok {
-				if k == "_id" { // if _id is not in projection map, do not do anything with it
-					continue
-				}
-				if inclusion {
-					docs[i].Remove(k)
-				}
+		err = projectDocument(inclusion, docs[i], projection)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func projectDocument(inclusion bool, doc *types.Document, projection *types.Document) error {
+	projectionMap := projection.Map()
+
+docMapLoopK1:
+	for k1 := range doc.Map() {
+		projectionVal, ok := projectionMap[k1]
+		if !ok {
+			if k1 == "_id" { // if _id is not in projection map, do not do anything with it
 				continue
 			}
+			if inclusion { // k1 from doc is absent in projection, remove from doc only if projection type inclusion
+				doc.Remove(k1)
+			}
+			continue
+		}
+
+		switch projectionVal := projectionVal.(type) { // found in the projection
+		case bool: // field: bool
+			if !projectionVal {
+				doc.Remove(k1)
+			}
+
+		case int32, int64, float64: // field: number
+			if compareScalars(projectionVal, int32(0)) == equal {
+				doc.Remove(k1)
+			}
+
+		case *types.Document: // field: { $elemMatch: { field2: value }}
+
+		procjectionDocLoop:
+			for _, projectionType := range projectionVal.Keys() {
+				supportedProjections := []string{"$elemMatch"}
+				if !slices.Contains(supportedProjections, projectionType) {
+					return fmt.Errorf("projecion %s is not supported", projectionType)
+				}
+
+				// for now it's only $elemMatch further
+				// if the corresponding value is not an array, skip
 
-			switch v := v.(type) { // found in the projection
-			case bool:
-				if !v {
-					docs[i].Remove(k)
+				docValueA, err := doc.GetByPath(k1)
+				if err != nil {
+					continue procjectionDocLoop
 				}
-			case int32, int64, float64:
-				if compareScalars(v, int32(0)) == equal {
-					docs[i].Remove(k)
+
+				// $elemMatch works only for arrays, it must be an array
+				docValueArray, ok := docValueA.(*types.Array)
+				if !ok {
+					doc.Remove(k1)
+					continue docMapLoopK1
+				}
+
+				// get the elemMatch conditions
+				conditions := must.NotFail(projectionVal.Get(projectionType)).(*types.Document)
+
+				for k2ConditionField, conditionValue := range conditions.Map() {
+					switch elemMatchFieldCondition := conditionValue.(type) {
+					case *types.Document: // TODO field2: { $gte: 10 }
+
+					case *types.Array:
+						panic("unexpected code")
+
+					default: // field2: value
+						found := -1 // >= 0 means found
+
+					internal:
+						for j := 0; j < docValueArray.Len(); j++ {
+							var cmpVal any
+							cmpVal, err = docValueArray.Get(j)
+							if err != nil {
+								continue internal
+							}
+							switch cmpVal := cmpVal.(type) {
+							case *types.Document:
+								docVal, err := cmpVal.Get(k2ConditionField)
+								if err != nil {
+									types.RemoveByPath(doc, k1, strconv.Itoa(j))
+									continue
+								}
+								if testutil.EqualScalars(docVal, elemMatchFieldCondition) {
```

Comment:
We definitly should not use testutil in non-testing code – that's a general-purpose functions. For example, it compares NaNs. All that is documented

Reply:
changed to compareScalars as agreed on the call.